### PR TITLE
SubscriptionStorage and Query Subscriptions async

### PR DESCRIPTION
**Pull request by danielmarbach**
This PR is part of the work to make the core async enabled. For more background read [Async/Await: It's time!](http://particular.net/blog/async-await-its-time) blogpost

@Particular/tf-asyncawait Please review

### CODE REVIEW

File: src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs [hunk 1]
```diff
@@ -55,8 +56,9 @@ public void Init()
             }
         }
 
-        public IEnumerable<string> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes)
+        public Task<IEnumerable<string>> GetSubscriberAddressesForMessage(IEnumerable<MessageType> messageTypes)
         {
+            var subscribers = new List<string>();
             var result = new List<string>();
             var messagelist = messageTypes.ToList();
             lock (locker)
```
Comment (SeanFeldman) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs]: @danielmarbach dumb question, but wouldn't this destroy async and block? Should we resort to an ["async lock"](http://blogs.msdn.com/b/pfxteam/archive/2012/02/12/10266988.aspx)?

Comment (SeanFeldman) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs]: @danielmarbach dumb question, but wouldn't this destroy async and block? Should we resort to an ["async lock"](http://blogs.msdn.com/b/pfxteam/archive/2012/02/12/10266988.aspx)?

Comment (danielmarbach) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs]: @SeanFeldman it is not a dumb question. It is an excellent question. Yes any type of lock would need an async lock. I considered the MsmqStorage something legacy, plus the underlying APIs aren't async in any way. Therefore I figured throwing in an async lock is pretty much worthless. @Particular/tf-asyncawait thoughts?

Comment (andreasohlund) [inline src/NServiceBus.Core/Persistence/Msmq/SubscriptionStorage/MsmqSubscriptionStorage.cs]: Doesn't seem like a big issue. That said we need better perf tests for this in the future

**Pull request merged**